crates/red_knot_project/src/walk.rs

```diff
@@ -0,0 +1,256 @@
+use crate::{Db, IOErrorDiagnostic, IOErrorKind, Project};
+use ruff_db::files::{system_path_to_file, File};
+use ruff_db::system::walk_directory::{ErrorKind, WalkDirectoryBuilder, WalkState};
+use ruff_db::system::{FileType, SystemPath, SystemPathBuf};
+use ruff_python_ast::PySourceType;
+use rustc_hash::{FxBuildHasher, FxHashSet};
+use std::path::PathBuf;
+use thiserror::Error;
+
+/// Filter that decides which files are included in the project.
+///
+/// In the future, this will hold a reference to the `include` and `exclude` pattern.
+///
+/// This struct mainly exists because `dyn Db` isn't `Send` or `Sync`, making it impossible
+/// to access fields from within the walker.
+#[derive(Default, Debug)]
+pub(crate) struct ProjectFilesFilter<'a> {
+    /// The same as [`Project::included_paths_or_root`].
+    included_paths: &'a [SystemPathBuf],
+
+    /// The filter skips checking if the path is in `included_paths` if set to `true`.
+    ///
+    /// Skipping this check is useful when the walker only walks over `included_paths`.
+    skip_included_paths: bool,
+}
+
+impl<'a> ProjectFilesFilter<'a> {
+    pub(crate) fn from_project(db: &'a dyn Db, project: Project) -> Self {
+        Self {
+            included_paths: project.included_paths_or_root(db),
+            skip_included_paths: false,
+        }
+    }
+
+    /// Returns `true` if a file is included according to the project's `include` and `exclude` settings
+    /// and the paths specified on the CLI.
+    ///
+    /// A file is considered being part of the project if it is a sub path of the project's root
```

Comment:
This suggests that "considered being part of the project" means the same thing as "included", but I find that confusing, because I would say if we have `project/a.py` and `project/b.py`, and we run `knot check project/a.py`, then only `a.py` is "included", but both of the files are still "part of the project", we just aren't checking all of them.
```suggestion
    /// A file is included if it is a sub path of the project's root
```